### PR TITLE
Loadout shares

### PR DESCRIPTION
**Pull request by bhollis**
This adds a new table for tracking "loadout shares", which are basically a full loadout keyed by a short ID. These IDs are somewhat naively generated as a random 32-bit integer, and checked for collisions before use. I'm keeping track of who uploaded each loadout, how many times it's been loaded, and when it was last loaded. I don't have plans to expose any of that info but I suspect it'll come in handy later, even just if we need to clean out old shares.

A new API `/loadout_share` can be used by authenticated users to post loadout shares. They get back a short URL like `http://dim.gg/oajjhqi/Foo-bar`.

At that URL is a server-rendered landing page which briefly summarizes the loadout and offers links to App and Beta. This allows us to change those loadout links over time - for example, to include more of the loadout, direct it to different pages, etc.

The page is very simple right now, because we don't have access to the manifest yet. I also haven't done any localization - I think we should just do something basic with i18next and the detected language from the request.

Screen Shot 2022-01-15 at 12 43 39 PM

### CODE REVIEW

File: api/dim-gg/views/loadout.ejs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      <p><%= loadout.notes %></p>
+      <% } %>
+      <p>
+        These build settings contain:
+        <ul>
+          <% if (numMods > 0) { %><li><%= numMods %> mods</li><% } %>
+          <% if (hasFashion) { %><li>Fashion (shaders &amp; ornaments)</li><% } %>
+          <% if (hasSubclass) { %><li>Stasis subclass settings</li><% } %>
+          <% if (hasLoParams) { %><li>Loadout Optimizer settings</li><% } %>
+          <% if (numMods > 0) { %><li><%= numMods %> mods</li><% } %>
```
Comment (kyleshay) [inline api/dim-gg/views/loadout.ejs:35]: FYI, mods listed twice (line 31 as well)

File: api/routes/loadout-share.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    res.status(400).send(validationResult);
+  }
+
+  const shareId = await transaction(async (client) => {
+    const attempts = 0;
+    // We'll make three attempts to guess a random non-colliding number
+    while (attempts < 4) {
+      const shareId = generateRandomShareId();
+      try {
+        addLoadoutShare(
```
Comment (kyleshay) [inline api/routes/loadout-share.ts:49]: noticed this is in an async fn, missing the `await`?

File: api/routes/loadout-share.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    if (validationResult) {
+      res.status(500).send({
+        status: 'RanOutOfIDs',
+        message: "We couldn't generate a share URL",
+      });
+      return;
+    }
+  }
+
+  const titleSlug = slugify(loadout.name);
```
Comment (kyleshay) [inline api/routes/loadout-share.ts:83]: is this mostly to just make the urls human readable/distinguishable?

Comment (bhollis) [reply api/routes/loadout-share.ts:83]: Yup, just for sharing

**Review by nev-r — APPROVED** [review]

**Pull request merged**